clang/include/clang/AST/ASTNodeTraverser.h

```diff
@@ -415,9 +443,55 @@ class ASTNodeTraverser
     if (!T->isSugared())
       Visit(T->getPattern());
   }
+  void VisitAutoType(const AutoType *T) {
+    for (const auto &Arg : T->getTypeConstraintArguments())
+      Visit(Arg);
+  }
   // FIXME: ElaboratedType, DependentNameType,
   // DependentTemplateSpecializationType, ObjCObjectType
 
+  // For TypeLocs, we automatically visit the inner type loc (pointee type etc).
+  // We must explicitly visit other lexically-nested nodes.
+  void VisitFunctionProtoTypeLoc(FunctionProtoTypeLoc TL) {
+    TypeLocVisitor<Derived>::VisitFunctionTypeLoc(TL);
+    for (const auto *Param : TL.getParams())
+      Visit(Param, /*VisitTypeLocs=*/true);
+  }
+  void VisitAutoTypeLoc(AutoTypeLoc TL) {
+    if (const auto *CR = TL.getConceptReference()) {
```

Comment:
And if it's not a concept reference? e.g., `void func(auto a) { decltype(auto) b = 12; }` (probably good test case to add)

Reply:
If this is not a concept reference there are no children to visit here.

Added tests for constrained & unconstrained AutoTypeLoc.
(`void func(auto a)` is not an AutoTypeLoc, it's a TemplateTypeParmTypeLoc)